patroni/quorum.py

```diff
@@ -0,0 +1,186 @@
+
+class QuorumError(Exception):
+    pass
+
+
+class QuorumStateResolver(object):
+    """
+    Calculates a list of state transition tuples of the form `('sync'/'quorum',number,set_of_names)`
+
+    Synchronous replication state is set in two places. PostgreSQL configuration sets how many and which nodes are
+    needed for a commit to succeed, abbreviated as `numsync` and `sync` set here. DCS contains information about how
+    many and which nodes need to be interrogated to be sure to see an xlog position containing latest confirmed commit,
+    abbreviated as `quorum` and `voters` set. Both pairs have the meaning "ANY n OF set".
+
+    The number of nodes needed for commit to succeed, `numsync`, is also called the replication factor.
+
+    To guarantee zero lost transactions on failover we need to keep the invariant that at all times any subset of
+    nodes that can acknowledge a commit overlaps with any subset of nodes that can achieve quorum to promote a new
+    leader. Given a desired replication factor and a set of nodes able to participate in sync replication there
+    is one optimal state satisfying this condition. Given the node set `active`, the optimal state is:
+
+        sync = voters = active
+        numsync = min(replication_factor, len(active))
+        quorum = len(active) + 1 - numsync
+
+    We need to be able to produce a series of state changes that take the system to this desired state from any
+    other state arbitrary given arbitrary changes is node availability, configuration and interrupted transitions.
+
+    To keep the invariant the rule to follow is that when increasing `numsync` or `quorum`, we need to perform the
+    increasing operation first. When decreasing either, the decreasing operation needs to be performed later.
+
+    For simplicity all sync members are considered equal. In Patroni the leader is actually special in that the last
+    known leader is always guaranteed to have latest state. This leads to suboptimal number of transitions when
+    increasing replication factor and quorum at the same time. If quorum must include leader, which happens when
+    transitioning to sync replication, i.e. `quorum, voters = 1, {'leader'}`, then the special leader semantics
+    mean that we could set `numsync, sync = 2, {'leader', 's1', 's2'}` in one step without worrying of the case that
+    s1, s2 have latest xlog, but not leader. The benefit seems too small to warrant adding any complexity.
+    """
+    def __init__(self, quorum, voters, numsync, sync, active, sync_wanted):
+        self.quorum = quorum
+        self.voters = set(voters)
+        self.numsync = numsync
+        self.sync = set(sync)
+        self.active = active
+        self.sync_wanted = sync_wanted
+
+    def check_invariants(self):
+        if self.quorum and not (len(self.voters | self.sync) < self.quorum + self.numsync):
+            raise QuorumError("Quorum and sync not guaranteed to overlap: nodes %d >= quorum %d + sync %d" %
```

Comment:
When I removed sync key from the DCS the master crashed:
```
$ etcdctl rm /service/batman/sync
PrevNode.Value: {"leader":"postgresql2","members":["postgresql1","postgresql0","postgresql2"],"quorum":2}

2019-02-10 09:51:33,719 INFO: Lock owner: postgresql2; I am postgresql2
2019-02-10 09:51:33.720 CET [10143] LOG:  received fast shutdown request
2019-02-10 09:51:33.725 CET [10143] LOG:  aborting any active transactions
2019-02-10 09:51:33.725 CET [10178] FATAL:  terminating connection due to administrator command
2019-02-10 09:51:33.730 CET [10143] LOG:  background worker "logical replication launcher" (PID 10193) exited with exit code 1
2019-02-10 09:51:33.731 CET [10146] LOG:  shutting down
2019-02-10 09:51:33.794 CET [10143] LOG:  database system is shut down
2019-02-10 09:51:33,813 INFO: Lock owner: postgresql2; I am postgresql2
Traceback (most recent call last):
  File "./patroni.py", line 6, in <module>
    main()
  File "/home/akukushkin/git/ants/patroni/patroni/__init__.py", line 182, in main
    return patroni_main()
  File "/home/akukushkin/git/ants/patroni/patroni/__init__.py", line 148, in patroni_main
    patroni.run()
  File "/home/akukushkin/git/ants/patroni/patroni/__init__.py", line 119, in run
    logger.info(self.ha.run_cycle())
  File "/home/akukushkin/git/ants/patroni/patroni/ha.py", line 1370, in run_cycle
    info = self._run_cycle()
  File "/home/akukushkin/git/ants/patroni/patroni/ha.py", line 1343, in _run_cycle
    msg = self.process_healthy_cluster()
  File "/home/akukushkin/git/ants/patroni/patroni/ha.py", line 966, in process_healthy_cluster
    'promoted self to leader because i had the session lock'
  File "/home/akukushkin/git/ants/patroni/patroni/ha.py", line 566, in enforce_master_role
    self.process_sync_replication()
  File "/home/akukushkin/git/ants/patroni/patroni/ha.py", line 426, in process_sync_replication
    sync_wanted=sync_wanted):
  File "/home/akukushkin/git/ants/patroni/patroni/quorum.py", line 84, in __iter__
    transitions = list(self._generate_transitions())
  File "/home/akukushkin/git/ants/patroni/patroni/quorum.py", line 95, in _generate_transitions
    self.check_invariants()
  File "/home/akukushkin/git/ants/patroni/patroni/quorum.py", line 62, in check_invariants
    (len(self.voters | self.sync), self.quorum, self.numsync))
patroni.quorum.QuorumError: Quorum and sync not guaranteed to overlap: nodes 3 >= quorum 1 + sync 2
```